c gen: 'main__func' undeclared

### Describe the bug

Wrong generated C code

```diff
VV_LOCAL_SYMBOL void main__main(void) {
	Map_string_main__func mapa = new_map_init(&map_hash_string, &map_eq_string, &map_clone_string, &map_free_string, 1, sizeof(string), sizeof(voidptr),
		_MOV((string[1]){
			_SLIT("la"),
		}),
		_MOV((voidptr[1]){
-			main__func, 
+			main__func_T_int, 
		})
	)
	;
	int val = 0;
	((int (*) (string arg, T* val))(*(voidptr*)map_get(ADDR(map, mapa), &(string[]){_SLIT("la")}, &(voidptr[]){ 0 })))(_SLIT("ran"), (voidptr)&/*qq*/val);
	 _v_dump_expr_int(_SLIT("a.v"), 14, _SLIT("func('la', mut val)"), main__func_T_int(_SLIT("la"), (voidptr)&/*qq*/val));
	// assert
	int _t1 = ((int (*) (string arg, T* val))(*(voidptr*)map_get(ADDR(map, mapa), &(string[]){_SLIT("la")}, &(voidptr[]){ 0 })))(_SLIT("ran"), (voidptr)&/*qq*/val);
	if (!(_t1 == 2)) {
		VAssertMetaInfo v_assert_meta_info__t2 = {0};
		v_assert_meta_info__t2.fpath = _SLIT("a.v");
		v_assert_meta_info__t2.line_nr = 15;
		v_assert_meta_info__t2.fn_name = _SLIT("main.main");
		v_assert_meta_info__t2.src = _SLIT("mapa['la']('ran', mut val) == 2");
		v_assert_meta_info__t2.op = _SLIT("==");
		v_assert_meta_info__t2.llabel = _SLIT("mapa['la']('ran', mut val)");
		v_assert_meta_info__t2.rlabel = _SLIT("2");
		v_assert_meta_info__t2.lvalue = int_str(_t1) /* typeof: v.ast.CTempVar type: ast.Type(0x8 = 8) */ ;
		v_assert_meta_info__t2.rvalue = string_clone(int_str(2)) /* typeof: v.ast.IntegerLiteral type: ast.Type(0x8 = 8) */ ;
		v_assert_meta_info__t2.has_msg = false;
		v_assert_meta_info__t2.message = _SLIT0;
		__print_assert_failure(&v_assert_meta_info__t2);
		// TODO
		// Maybe print all vars in a test function if it fails?
		_v_panic(_SLIT("Assertion failed..."));
	}
}
```

### Reproduction Steps

```v
fn func[T](arg string, mut val T) int {
	println('hello from func')
	return 2
}

fn main() {

	mapa := {"la": func}

	mut val := 0

	mapa['la']('ran', mut val)

	dump(func('la', mut val))

	assert mapa['la']('ran', mut val) == 2
}
```

### Expected Behavior

assert pass

### Current Behavior

```sh
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/a.01JBN9F18JQDN28NF2SJ0593VJ.tmp.c:13629: error: 'main__func' undeclared
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

latest

### Environment details (OS name and version, etc.)

linux

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzI1ODg3ZDI3Y2M3NjgxYzYyNWQ2ZWQiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.O2Ks7qNcnTZwdoBQo-_YVdVe1z-7ZwbLxWe4gEPDvEg">Huly&reg;: <b>V_0.6-21180</b></a></sub>

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3833,6 +3833,10 @@ fn (mut c Checker) ident(mut node ast.Ident) ast.Type {
 		info := node.info as ast.IdentFn
 		if func := c.table.find_fn(node.name) {
 			if func.generic_names.len > 0 {
+				if node.concrete_types.len == 0 {
+					c.error('`${node.name}` is a generic fn, you should pass its concrete types, e.g. ${node.name}[int]',
+						node.pos)
+				}
 				concrete_types := node.concrete_types.map(c.unwrap_generic(it))
 				if concrete_types.all(!it.has_flag(.generic)) {
 					c.table.register_fn_concrete_types(func.fkey(), concrete_types)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -4010,6 +4010,12 @@ fn (mut c Checker) ident(mut node ast.Ident) ast.Type {
 		}
 		// Non-anon-function object (not a call), e.g. `onclick(my_click)`
 		if func := c.table.find_fn(name) {
+			if func.generic_names.len > 0 {
+				if node.concrete_types.len == 0 {
+					c.error('`${node.name}` is a generic fn, you should pass its concrete types, e.g. ${node.name}[int]',
+						node.pos)
+				}
+			}
 			return c.resolve_var_fn(func, mut node, name)
 		}
 	}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -121,8 +121,9 @@ mut:
 	is_last_stmt                     bool
 	prevent_sum_type_unwrapping_once bool // needed for assign new values to sum type, stopping unwrapping then
 	using_new_err_struct             bool
-	need_recheck_generic_fns         bool // need recheck generic fns because there are cascaded nested generic fn
-	inside_sql                       bool // to handle sql table fields pseudo variables
+	need_recheck_generic_fns         bool            // need recheck generic fns because there are cascaded nested generic fn
+	generic_fns                      map[string]bool // register generic fns that needs recheck once
+	inside_sql                       bool            // to handle sql table fields pseudo variables
 	inside_selector_expr             bool
 	inside_interface_deref           bool
 	inside_decl_rhs                  bool

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -38,9 +38,10 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 		// This is done so that all generic function calls can
 		// have a chance to populate c.table.fn_generic_types with
 		// the correct concrete types.
-		if !c.generic_fns[node.fkey()] {
+		fkey := node.fkey()
+		if !c.generic_fns[fkey] {
 			c.need_recheck_generic_fns = true
-			c.generic_fns[node.fkey()] = true
+			c.generic_fns[fkey] = true
 			c.file.generic_fns << node
 		}
 		return

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -38,8 +38,11 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 		// This is done so that all generic function calls can
 		// have a chance to populate c.table.fn_generic_types with
 		// the correct concrete types.
-		c.file.generic_fns << node
-		c.need_recheck_generic_fns = true
+		if !c.generic_fns[node.fkey()] {
+			c.need_recheck_generic_fns = true
+			c.generic_fns[node.fkey()] = true
+			c.file.generic_fns << node
+		}
 		return
 	}
 	node.ninstances++
@@ -2764,7 +2767,7 @@ fn (mut c Checker) post_process_generic_fns() ! {
 		fkey := node.fkey()
 		all_generic_fns[fkey]++
 		if all_generic_fns[fkey] > generic_fn_cutoff_limit_per_fn {
-			c.error('generic function visited more than ${generic_fn_cutoff_limit_per_fn} times',
+			c.error('${fkey} generic function visited more than ${generic_fn_cutoff_limit_per_fn} times',
 				node.pos)
 			return error('fkey: ${fkey}')
 		}

--- a/vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.out
+++ b/vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.out
@@ -5,6 +5,20 @@ vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:2: warning: unused 
       |     ~~~
     7 |     // ff1 := f1[int] <-- is a valid usage with generic return types that are not generic structs
     8 | }
+vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:9: error: `f1` is a generic fn, you should pass its concrete types, e.g. f1[int]
+    4 | 
+    5 | fn main() {
+    6 |     ff1 := f1 // <-- missing e.g. `[int]`
+      |            ~~
+    7 |     // ff1 := f1[int] <-- is a valid usage with generic return types that are not generic structs
+    8 | }
+vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:9: error: `f1` is a generic fn, you should pass its concrete types, e.g. f1[int]
+    4 | 
+    5 | fn main() {
+    6 |     ff1 := f1 // <-- missing e.g. `[int]`
+      |            ~~
+    7 |     // ff1 := f1[int] <-- is a valid usage with generic return types that are not generic structs
+    8 | }
 vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:9: error: cannot assign `f1` as a generic function variable
     4 | 
     5 | fn main() {

--- a/vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.out
+++ b/vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.out
@@ -26,8 +26,3 @@ vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:6:9: error: cannot as
       |            ~~
     7 |     // ff1 := f1[int] <-- is a valid usage with generic return types that are not generic structs
     8 | }
-vlib/v/checker/tests/generic_fn_infinite_loop_limit_err.vv:1:1: error: generic function visited more than 10000 times
-    1 | fn f1[T](x T, i int) T {
-      | ~~~~~~~~~~~~~~~~~~~~~~
-    2 |     return x
-    3 | }

--- a/vlib/v/checker/tests/generic_fn_var_unresolved_err.out
+++ b/vlib/v/checker/tests/generic_fn_var_unresolved_err.out
@@ -19,10 +19,3 @@ vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:10:11: error: `func` is a 
       |                 ~~~~
    11 |     }
    12 | }
-vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:3:1: error: generic function visited more than 10000 times
-    1 | module main
-    2 | 
-    3 | fn func[T](arg string, mut val T) int {
-      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-    4 |     return 2
-    5 | }

--- a/vlib/v/checker/tests/generic_fn_var_unresolved_err.out
+++ b/vlib/v/checker/tests/generic_fn_var_unresolved_err.out
@@ -1,0 +1,13 @@
+vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:8:7: error: `func` is a generic fn, you must pass its concrete types, e.g. func[int]
+    6 | 
+    7 | fn main() {
+    8 |     _ := func
+      |          ~~~~
+    9 | }
+vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:3:1: error: generic function visited more than 10000 times
+    1 | module main
+    2 | 
+    3 | fn func[T](arg string, mut val T) int {
+      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    4 |     return 2
+    5 | }

--- a/vlib/v/checker/tests/generic_fn_var_unresolved_err.out
+++ b/vlib/v/checker/tests/generic_fn_var_unresolved_err.out
@@ -1,9 +1,24 @@
-vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:8:7: error: `func` is a generic fn, you must pass its concrete types, e.g. func[int]
+vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:8:7: error: `func` is a generic fn, you should pass its concrete types, e.g. func[int]
     6 | 
     7 | fn main() {
     8 |     _ := func
       |          ~~~~
-    9 | }
+    9 |     _ := {
+   10 |         'test': func
+vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:8:7: error: `func` is a generic fn, you should pass its concrete types, e.g. func[int]
+    6 | 
+    7 | fn main() {
+    8 |     _ := func
+      |          ~~~~
+    9 |     _ := {
+   10 |         'test': func
+vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:10:11: error: `func` is a generic fn, you should pass its concrete types, e.g. func[int]
+    8 |     _ := func
+    9 |     _ := {
+   10 |         'test': func
+      |                 ~~~~
+   11 |     }
+   12 | }
 vlib/v/checker/tests/generic_fn_var_unresolved_err.vv:3:1: error: generic function visited more than 10000 times
     1 | module main
     2 | 

--- a/vlib/v/checker/tests/generic_fn_var_unresolved_err.vv
+++ b/vlib/v/checker/tests/generic_fn_var_unresolved_err.vv
@@ -6,4 +6,7 @@ fn func[T](arg string, mut val T) int {
 
 fn main() {
 	_ := func
+	_ := {
+		'test': func
+	}
 }

--- a/vlib/v/checker/tests/generic_fn_var_unresolved_err.vv
+++ b/vlib/v/checker/tests/generic_fn_var_unresolved_err.vv
@@ -1,0 +1,9 @@
+module main
+
+fn func[T](arg string, mut val T) int {
+	return 2
+}
+
+fn main() {
+	_ := func
+}

